src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,212 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
+            private readonly IEnumerable<TSource> _third;
+
+            internal Concat3Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second, IEnumerable<TSource> third)
+            {
+                Debug.Assert(first != null && second != null && third != null);
+                _first = first;
+                _second = second;
+                _third = third;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new Concat3Iterator<TSource>(_first, _second, _third);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                return new ConcatNIterator<TSource>(this, next, 3);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    case 2: return _third;
+                    default: return null;
+                }
+            }
+        }
+
+        private sealed class ConcatNIterator<TSource> : ConcatIterator<TSource>
+        {
+            // To handle chains of >= 4 sources, we chain the concat iterators together and allow
+            // GetEnumerable to fetch enumerables from the previous sources.  This means that rather
+            // than each MoveNext/Current calls having to traverse all of the previous sources, we
+            // only have to traverse all of the previous sources once per chained enumerable.  An
+            // alternative would be to use an array to store all of the enumerables, but this has
+            // a much better memory profile and without much additional run-time cost.
+
+            private readonly ConcatIterator<TSource> _previousConcat;
+            private readonly IEnumerable<TSource> _next;
+            private readonly int _nextIndex;
+
+            internal ConcatNIterator(ConcatIterator<TSource> previousConcat, IEnumerable<TSource> next, int nextIndex)
+            {
+                Debug.Assert(previousConcat != null);
+                Debug.Assert(next != null);
+                Debug.Assert(nextIndex > 0);
+                _previousConcat = previousConcat;
+                _next = next;
+                _nextIndex = nextIndex;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNIterator<TSource>(_previousConcat, _next, _nextIndex);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                return new ConcatNIterator<TSource>(this, next, checked(_nextIndex + 1));
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                return
+                    index < _nextIndex ? _previousConcat.GetEnumerable(index) :
```

Comment:
I'm not against that, but this was by far the simplest mechanism I could come up with, and the cost here should in general be very minimal.  Since we'd need to process from the oldest to the newest (which is the opposite direction of the chain), and since we can't rewrite the chain, how would you recommend doing this without manually building up a stack (which has its own costs)?